### ISSUE TITLE
SES on webasm

### ISSUE THREAD

**Issue by dckc** [Agoric/SES#19]
@dtribble suggested using XS, but since it throws away the source code of functions, I can't imagine it will work. I looked at otto and its successor https://github.com/dop251/goja , but it's not quite clear how to get started with those.

So I tried https://duktape.org/ , written in C. Compiling it to webasm was entirely straightforward:

After [installing emcc](https://kripken.github.io/emscripten-site/docs/getting_started/downloads.html
):

```
# How does this thing normally compile?
~/projects/duktape-2.3.0$ make -f Makefile.cmdline 
gcc -o duk  -Os -pedantic -std=c99 -Wall -fstrict-aliasing -fomit-frame-pointer -I./examples/cmdline -I./src    -DDUK_CMDLINE_PRINTALERT_SUPPORT -I./extras/print-alert -DDUK_CMDLINE_CONSOLE_SUPPORT -I./extras/console -DDUK_CMDLINE_LOGGING_SUPPORT -I./extras/logging -DDUK_CMDLINE_MODULE_SUPPORT -I./extras/module-duktape src/duktape.c examples/cmdline/duk_cmdline.c extras/print-alert/duk_print_alert.c extras/console/duk_console.c extras/logging/duk_logging.c extras/module-duktape/duk_module_duktape.c -lm

# Let's throw out a few options and feed the rest to emcc...
connolly@jambox:~/projects/duktape-2.3.0$ emcc -s WASM=1 -o duk.html -I./examples/cmdline -I./src    -DDUK_CMDLINE_PRINTALERT_SUPPORT -I./extras/print-alert -DDUK_CMDLINE_CONSOLE_SUPPORT -I./extras/console -DDUK_CMDLINE_LOGGING_SUPPORT -I./extras/logging -DDUK_CMDLINE_MODULE_SUPPORT -I./extras/module-duktape src/duktape.c examples/cmdline/duk_cmdline.c extras/print-alert/duk_print_alert.c extras/console/duk_console.c extras/logging/duk_logging.c extras/module-duktape/duk_module_duktape.c -lm
INFO:root:generating system asset: is_vanilla.txt... (this will be cached in "/home/connolly/.emscripten_cache/is_vanilla.txt" for subsequent builds)
...
connolly@jambox:~/projects/duktape-2.3.0$ firefox duk.html
```

And voila, a js repl in webasm running in firefox.

According to https://duktape.org/guide.html , const has function scope and there's no support for ES6 modules. But they talk about hooking babel into their module support... maybe it could make up the difference? The TCB is starting to get uncomfortably large at that point, I suppose.

**Comment by indolering** [comment]
Is there discussion somewhere about why it would be desirable to include a JS interpreter in SES?  The drawbacks are severe.

**Comment by dckc** [comment]
Did you mean to raise a separate issue? I don't see how the question is relevant to SES on webasm.

**Comment by indolering** [comment]
> Did you mean to raise a separate issue? I don't see how the question is relevant to SES on webasm.

I was going to comment on the feasibility of the task, but if I don't know the end goal I can't really comment on how you can meet those requirements.  Is there a Matrix/IRC/Slack channel that SES developers idle on?

**Comment by dckc** [comment]
feasibility of which task? SES on webasm or a JS interpreter in SES? They seem like completely different things, to me.

I think the motivation for SES on webasm is that Ethereum seems to be heading toward webasm support, so if we had an SES interepreter in webasm, we could run SES on Ethereum.

SES expects `eval()` as a standard feature, so I don't see motivation for a JS interpreter in SES.

As to chat channels... the #erights channel on irc.freenode.net is home to all sorts of discussion of object capability systems, but I don't think the SES developers participate directly.

**Comment by erights** [comment]
There are two reasons we're interested in running SES on XS:
   * For uses of SES that require high confidence in determinism, correctness, and/or security, we want to run on an implementation of JS that's simple enough to understand and make reliable. This disqualifies any of the current JIT-based JS engines built by browser makers, which are optimized for speed at the cost of complexity. By contrast, XS is an interpreter optimized for small footprint, for use on small memory devices. These pressures lead towards the kind of simplicity needed as a first step towards reliability.
   * wasm has emerged as a standard target for running general purpose computations on blockchains. Last I counted, at least six prominent blockchain projects were either already committed to wasm or were seriously trying to transition to it. XS is written in C/C++. There is a good C/C++ compiler and associated toolchain targeting wasm. We hope to compile XS to wasm.

**Comment by indolering** [comment]
> SES expects eval() as a standard feature, so I don't see motivation for a JS interpreter in SES.

Yes, it doesn't make sense at all ... which is why I was so confused.

> For uses of SES that require high confidence in determinism, correctness, and/or security, we want to run on an implementation of JS that's simple enough to understand and make reliable.

On the off-chance that this is helpful: the K Framework team has a [fancy new LLVM based compiler](https://runtimeverification.com/blog/new-technologies-for-the-blockchain-iele-virtual-machine-and-k-universal-language-framework/) which can extract competitive interpreters from formal language specifications.  However, their [JavaScript specification](https://github.com/kframework/javascript-semantics) is out of date, does not support the standard library, and I have no idea what the output size is like.

> As to chat channels... the #erights channel on irc.freenode.net is home to all sorts of discussion of object capability systems, but I don't think the SES developers participate directly.

That's unfortunate, I was thinking of clearing some low-hanging fruit and real-time communication would be helpful.

I'll try not to spam your tickets 😃 .

**Comment by erights** [comment]
(Sorry, closed by mistake. Reopened.)

**Comment by jfparadis** [comment]
We're discussing this today! It's now back on my front burner.
